### PR TITLE
require special permission for cluster building

### PR DESCRIPTION
**Pull request by wolfreak100**
Because I do not know too much about game_do_command yet, i did not want
to modify it at the expense of introducing bugs. This currently serves
no purpose other than to allow a game command function, as the
functionality uses network_can_do_action instead. It's a little hackish,
but it gets the job done for right now.

### CODE REVIEW

**Comment by zsilencer** [comment]
You can just do the same thing as the chat permission, which is using -1, use -2, no need to add a game command.

File: data/language/english_uk.txt [hunk 1]
```diff
@@ -4003,6 +4003,7 @@ STR_5661    :Set Player Group
 STR_5662    :N/A
 STR_5663    :Clear Landscape
 STR_5664    :Cheat
+STR_5665	:Toggle Scenery Cluster
```
Comment (IntelOrca) [inline data/language/english_uk.txt]: Use spaces, not tabs here... you can see the alignment is wrong.

File: src/localisation/string_ids.h [hunk 2]
```diff
@@ -2303,7 +2303,7 @@ enum {
 	STR_ACTION_NA = 5662,
 	STR_ACTION_CLEAR_LANDSCAPE = 5663,
 	STR_ACTION_CHEAT = 5664,
-	
+	STR_ACTION_TOGGLE_SCENERY_CLUSTER = 5665,
```
Comment (LinusU) [inline src/localisation/string_ids.h]: You should probably keep an empty line here since it was made to divide the constants

File: src/network/network.cpp [hunk 3]
```diff
@@ -2407,7 +2407,9 @@ int network_can_perform_action(unsigned int groupindex, unsigned int index)
 {
 	return gNetwork.group_list[groupindex]->CanPerformAction(index);
 }
-
```
Comment (LinusU) [inline src/network/network.cpp]: Same here, better to keep the empty line to follow the existing style

File: src/windows/scenery.c [hunk 4]
```diff
@@ -33,7 +33,7 @@
 #include "../world/sprite.h"
 #include "dropdown.h"
 #include "../interface/themes.h"
-
+#include "../network/network.h"
```
Comment (LinusU) [inline src/windows/scenery.c]: Same here, an empty line between the `include`s and `define`s helps with readability

File: src/world/scenery.h [hunk 5]
```diff
@@ -187,5 +187,5 @@ void scenery_update_age(int x, int y, rct_map_element *mapElement);
 void scenery_set_default_placement_configuration();
 void scenery_remove_ghost_tool_placement();
 void scenery_set_default_placement_configuration();
-
+int game_command_toggle_scenery_cluster(int* eax, int* ebx, int* ecx, int* edx, int* esi, int* edi, int* ebp);
```
Comment (LinusU) [inline src/world/scenery.h]: An empty line before the `endif` would look good

File: src/world/scenery.c [hunk 6]
```diff
@@ -193,3 +193,7 @@ void scenery_remove_ghost_tool_placement(){
 			0);
 	}
 }
+int game_command_toggle_scenery_cluster(int* eax, int* ebx, int* ecx, int* edx, int* esi, int* edi, int* ebp) {
+	//Because I do not know too much about game_do_command yet, i did not want to modify it at the expense of introducing bugs. This currently serves no purpose other than to allow a game command function, as the functionality uses network_can_do_action instead. It's a little hackish, but it gets the job done for right now. -wolfreak_99
+	return 1;
+}
```
Comment (LinusU) [inline src/world/scenery.c]: Missing newline at end of file, should probably have an empty line above the function as well

File: src/windows/top_toolbar.c [hunk 7]
```diff
@@ -1461,10 +1462,11 @@ static void window_top_toolbar_scenery_tool_down(short x, short y, rct_window *w
 	case SCENERY_TYPE_SMALL:
 	{
 		int quantity = 1;
-		if (window_scenery_is_build_cluster_tool_on) {
-			quantity = 35;
+		bool isCluster = window_scenery_is_build_cluster_tool_on && network_can_perform_action(network_get_current_player_group_index(), GAME_COMMAND_TOGGLE_SCENERY_CLUSTER);
+		if (isCluster) {
+			quantity = 36;
```
Comment (LinusU) [inline src/windows/top_toolbar.c]: Why did you increment `quantity` to 36?

File: src/windows/top_toolbar.c [hunk 8]
```diff
@@ -1461,10 +1462,11 @@ static void window_top_toolbar_scenery_tool_down(short x, short y, rct_window *w
...
 		}
-
+
```
Comment (LinusU) [inline src/windows/top_toolbar.c]: You added a bunch of spaces here

File: src/windows/top_toolbar.c (hunk 7 above)
Comment (wolfreak99) [inline src/windows/top_toolbar.c]: That was by accident, i thought i read it as 35.  Like, i have a forked version of the code that contains a bunch of experimental features (autosave enable/disable checkbox in file menu, private messaging, customizable kicking messages, object selection modified to work in multiplayer, logging players IP and name when they join, a layout designer for designing and layout out widgets for windows) that contains a "_project_flags.h" file with a bunch of defines to enable/disable/isolate features i add in. When I submitted this code I created the wolfreak100 account to have 2 forks of the project, that way i could implement this feature only and send it, so when i copied the parts over i must have overlooked 35 as 36.

I've been meaning to go back and edit it to revise it, but lately i noticed the slowdown of accepting of pull requests so i've not been in really much of a rush tbh.

Comment (IntelOrca) [inline src/windows/top_toolbar.c]: @wolfreak99 we would like to merge this in for 0.0.4 due to the number of complains about abusers of this. Have you tried implementing it like @zsilencer suggested?

Comment (wolfreak100) [inline src/windows/top_toolbar.c]: @IntelOrca  forreal, bro? I'll restart into windows 8 now and try getting it in there then. Outta curiosity, are you going to put that disable litter in as well?

I also have a question, as I'm new with c a bit and i'm trying to understand the concept of the networking, could you explain how it works? What are the connection steps done with networkconnection, networkaddress, networkplayer, etc?

Comment (janisozaur) [inline src/windows/top_toolbar.c]: @wolfreak100 are you @wolfreak99 ? why the number change?

Comment (wolfreak101) [inline src/windows/top_toolbar.c]: @janisozaur Because you can be anyone you want to be on the internet.

Comment (wolfreak100) [inline src/windows/top_toolbar.c]: @janisozaur  but to be real though, i did it because i have 2 forks of this project. one for experimenting with, and two for the push i did of this. I could't figure out ho to get 2 forks of the code without having to spend alot of time and effort. 

@IntelOrca also, i'm looking at this code of the chat command, and honestly it's not making much sense to me on how it works. it's doing a "packet << (packet)chat_command;" (or something along that line), and it's just, i'm gonna have to study how this chat command works which would take some time. and honestly right now... sure it's a little hacky but hey, the code works. plus openrct2 is not even in an alpha state. i can switch this up later when i get time but im just busy with alot of other stuff as of the last few days, and my current goal openrct2-relatd is figuring out why a window i made is skipping a widget index in mousedown (it's reading me pressing widget 4 when i'm clicking over widget 3, but likely i did something wrong in the code that i just don't see) 

i'm sure right now the multiplayer community would care more about getting a feature that might not be properly implemented right under the hood than not get it at all, right? it's like indie game developing, in the end the person playing the game doesn't care if you use unity, gamemaker, or unreal, if it's a good game they'll like it. 

if you want to take this idea and adjust it to how silencer suggested if you think it would be easy for you, go ahead and take my work and recreate it yourself, you don't even need to credit me :)

Comment (zsilencer) [inline src/windows/top_toolbar.c]: all you have to do is get rid of the game command and make STR_ACTION_TOGGLE_SCENERY_CLUSTER = {-2}, and use -2 in network_can_perform_action, look at STR_ACTION_CHAT

**Comment by dlee13** [comment]
@wolfreak100 No need to fork the project twice, bro. Use branches. You need to learn Git if you want to collaborate on a project.

**Comment by wolfreak99** [comment]
@dlee13 would I be able to create a separate branch from my own forked version, and can I do it through the github's web/Windows software/visual studio plugins GUI? or do I need to learn alot of complicated shell commands? I did a separate account cause it seemed the quickest and less-hassle solution at the time, since all I was gonna do is copy paste my edits into the repository and that would have been it. If you couldhelp me though I'd be willing to learn, shoot me an email . jbminor1991@gmail.com

**Comment by Sixmorphugus** [comment]
@wolfreak99 this is actually a surprisingly hard thing to do, so I'll walk you through it.

You're working in visual studio (i gather) so you need to go into "changes" in the team collaboration menu, then `Actions >> Open Command Prompt`.

Type:
`git branch [branchname]` to create a new branch (note that this does NOT publish a new branch)

`git checkout [branchname]` to switch to the new branch (which will be in the same state as the branch your repository was in when you created the new branch)

`git remote add upstream https://github.com/OpenRCT2/OpenRCT2` will just make things easier in the long run if upstream doesn't already exist. Upstream being the remote branch you're working from - i.e. `OpenRCT2/develop`

`git fetch upstream` will check the current state of the remote branch and store it in the working tree or something. You need to type it for the next command to work, and that's all I know.

`git reset --hard upstream/develop` will ERASE all of the edits made to the local branch you have checked out and basically rewrite history to synchronize you with the official OpenRCT2 branch and erase your previous experiments. Be CAREFUL with this - be sure your other changes are saved in the other branch.

Finally,
`git push -u origin [branchname]` will publish the new branch to your account and you can start working on new changes without having lost anything or having to fork twice. The local repo will now track this online branch as its remote.

**Comment by janisozaur** [comment]
@wolfreak100 @wolfreak101 @wolfreak99 or whoever you are, based on your comment https://github.com/OpenRCT2/OpenRCT2/pull/3241#issuecomment-205169016 are we to understand you will **not** be working on this code?

**Comment by wolfreak100** [comment]
I've updated the code. I apologize about my recent attitude. everything in my life has been rough lately.

Not sure how to push the latest commit to this yet.

File: src/localisation/string_ids.h (hunk 2 above)
Comment (Gymnasiast) [inline src/localisation/string_ids.h]: Indeed.

File: src/windows/top_toolbar.c [hunk 9]
```diff
@@ -1461,10 +1462,11 @@ static void window_top_toolbar_scenery_tool_down(short x, short y, rct_window *w
 	case SCENERY_TYPE_SMALL:
 	{
 		int quantity = 1;
-		if (window_scenery_is_build_cluster_tool_on) {
+		bool isCluster = window_scenery_is_build_cluster_tool_on && (network_get_mode() != NETWORK_MODE_CLIENT || network_can_perform_action(network_get_current_player_group_index(), -2));
+		if (isCluster) {
 			quantity = 35;
 		}
-
+
```
Comment (Gymnasiast) [inline src/windows/top_toolbar.c]: Could you remove this trailing whitespace?

File: src/world/scenery.h [hunk 10]
```diff
@@ -187,5 +187,4 @@ void scenery_update_age(int x, int y, rct_map_element *mapElement);
 void scenery_set_default_placement_configuration();
 void scenery_remove_ghost_tool_placement();
 void scenery_set_default_placement_configuration();
-
```
Comment (Gymnasiast) [inline src/world/scenery.h]: I'd reinstate this empty line.

File: src/network/network.h [hunk 11]
```diff
@@ -202,7 +203,7 @@ class NetworkActions
 		{STR_ACTION_KICK_PLAYER, {GAME_COMMAND_KICK_PLAYER}},
 		{STR_ACTION_MODIFY_GROUPS, {GAME_COMMAND_MODIFY_GROUPS}},
 		{STR_ACTION_SET_PLAYER_GROUP, {GAME_COMMAND_SET_PLAYER_GROUP}},
-		{STR_ACTION_CHEAT, {GAME_COMMAND_CHEAT}}
+		{STR_ACTION_CHEAT, {GAME_COMMAND_CHEAT}},
```
Comment (IntelOrca) [inline src/network/network.h]: Unnecessary line change

**Comment by wolfreak99** [comment]
Okay. I'm at work at the moment, but when I go home I will be able to get these formatting issues fixed.. If you could go through and find any issues and let me know of them (including the ones before, as I don't remember what all I've changed since I updated this..), i will address them when i get home. If i do will this be merged? Or if not how long would it take? I'm just asking cause I don't wanna be stuck in limbo lol :P



Also, intelorca, i have created an ip based banning system, that isnt a permanent banning solution, its an increment based banning system. It has 30 seconds, 2 minutes, 5 minutes, 15 minutes, 30 minutes,  1 hour, 2 hours, 3 hours, 6 hours , 12 hours, 1 day, 2 days, 1 week, 2 weeks, never expires, end of host session. There are predefined reasons for banning people such as "vandalism" "spamming" "hate speech" "inappropriate content" "illegal content" "circumventing ban" "other reason", the other reason brings up a box where you can type the reason in. It works but i was also learning c at the time of dojng it so i am gojng to use this repo to reimplement it correctly. I've worked on it for about a month now. I just wanna know will there be a chance of you committing it? I will possibly throw in some work to combine a systems info into a hash of sorts, but for that to fully work, it will have to be in the clients code as well, meaning I want confirmation that there's at least a chance of it being pushed before i take more time (remember im ralking a month on this, and longer just to figure out ip address retrieving) to add a feature that clients will need to have (right nkw, my ip banning works fine because it detects it server side as well).. I hope I don't sound harsh, I'm genuinely curious about this. I know you guys are busy with decomping the game and adding the harscore code, that's why I'm wanting to offer my chances of improving the multiplayer moderation tools. I also have a prototype for private messaging as well, but names are required to be alphanumeric, and contain only _-' (apostrophe included for foreign names like Ra'quel-son or.. something), the name guideline I use is it must be 2-18 chars, begin and end with a letter or number, and no trailing symbols such as joe__man. The private messaging currently works like irc would "/p IntelOrca hey bro" would send you a message and if you wrote "/r", you'd get "/p wolfreak99 " put into your chat input section for quicker replying.. I'm just curious on all of this.

**Comment by IntelOrca** [comment]
This no longer works.

CanPerformAction(-2), -2 gets casted to unsigned and then fails. If you look at chat, that is using ```CanPerformCommand``` and inside that function is has a special case for -1, so you will likely have to add a special case for -2.

You should also have the cluster building permission off by default. Currently its on for spectators.

**Comment by wolfreak100** [comment]
@IntelOrca  How would I go about turning it off for spectators by default? I also just realized that, by having the index for it -2 now, it will be above chat. the group configurations file is set to load from 0 below.. Will there be a way to make Toggle Scenery Cluster be at the very bottom of the list again, like it was back when i had it using the blank gamecommand function? I don't want the community's group permissons file to be completely offset :(. Like, this is really scary now. I'm trying to help the community, not sabbotage it even more...

**Comment by IntelOrca** [comment]
@wolfreak100 Move ```[STR_ACTION_TOGGLE_SCENERY_CLUSTER, -2]``` to the bottom of ```actions```.

File: src/network/network.cpp [hunk 12]
```diff
@@ -232,8 +232,8 @@ int NetworkActions::FindCommand(int command)
 	if (it != actions.end()) {
 		return it - actions.begin();
 	}
-	if (command < 0) {
-		return command;
+	if (command == -2) {
+		return -2;
 	}
```
Comment (IntelOrca) [inline src/network/network.cpp]: This is unnecessary.

File: src/network/network.cpp [hunk 13]
```diff
@@ -232,6 +232,9 @@ int NetworkActions::FindCommand(int command)
 	if (it != actions.end()) {
 		return it - actions.begin();
 	}
+	if (command == -2) {
+		return -2;
+	}
```
Comment (IntelOrca) [inline src/network/network.cpp]: This is unnecessary.

File: src/network/network.h [hunk 14]
```diff
@@ -184,7 +184,7 @@ class NetworkActions
 public:
 	int FindCommand(int command);
 	const std::vector<NetworkAction> actions = {
-		{STR_ACTION_CHAT, {-1}},
+		{ STR_ACTION_CHAT,{ -1 } },
```
Comment (IntelOrca) [inline src/network/network.h]: Revert this line change.

File: src/world/scenery.h [hunk 15]
```diff
@@ -188,4 +188,4 @@ void scenery_set_default_placement_configuration();
 void scenery_remove_ghost_tool_placement();
 void scenery_set_default_placement_configuration();
 
-#endif
+#endif
```
Comment (IntelOrca) [inline src/world/scenery.h]: Revert this change.

**Comment by wolfreak100** [comment]
wait. whats this travis ci build about? does this mean it's not getting merged?

**Comment by janisozaur** [comment]
It does not compile with network disabled.

**Comment by wolfreak99** [comment]
Would this be a barrier to it not getting merged? Like.. I already started working on readding the messaging commands into it. I created a seperate branch based off the last commit on here and started working on adding the messaging and chat commands, but I'm still very new to branches and not sure if I'll have access to this code. I mean I guess I could download the zip to the code and rebuild it and write down any changes I do and then use the online github editor to insert the changes, it'll suck but this is so close to being finished, and I know the community needs it soon, so I feel kind of a responsibility to work to get this merged... How would I go about disabling the network so I can find out any errors causing it? I use visual studio. I know it has something to do with the flag ifdef NETWORK_DISABLED, but I'm not sure how to flag it with the network disabled

**Comment by zsilencer** [comment]
You can see where travis-ci is failing, you need to add the network_can_perform_command function for when network is disabled like you did with your other new function.  This cluster scenery issue seems like it would be simpler and easier to solve by just moving the cluster building into the place scenery game command and passing it a flag.  Abuse can be prevented by only allowing the admins (group id 0) to use cluster scenery because how often is this needed in a normal game anyway?

**Comment by Gymnasiast** [comment]
In the end, we decided to take your work and implement the missing parts ourselves, so the users don't have to miss out. Thank you for your contribution.
